WcaOnRails/lib/paypal_interface.rb

```diff
@@ -0,0 +1,128 @@
+                },
+              },
+            },
+          },
+        ],
+        products: ['PPCP'], # TODO: Experiment with other payment types
+        partner_config_override: {
+          return_url: EnvConfig.ROOT_URL + Rails.application.routes.url_helpers.competitions_paypal_return_path(competition_id),
+          return_url_description: "the url to return the WCA after the paypal onboarding process.",
+        },
+        legal_consents: [
+          {
+            type: 'SHARE_DATA_CONSENT',
+            granted: true,
+          },
+        ],
+      }
+
+      response = paypal_connection(url).post do |req|
+        req.body = payload
+      end
+
+      response.body['links'].each do |link|
+        if link['rel'] == "action_url"
+          return link['href']
+        end
+      end
+    end
+  end
+
+  def self.create_order(registration)
+    unless Rails.env.production?
+      url = "#{EnvConfig.PAYPAL_BASE_URL}/v2/checkout/orders"
+
+      outstanding_fees = registration.outstanding_entry_fees
+      fee_currency = outstanding_fees.currency.iso_code
+
+      payload = {
+        intent: 'CAPTURE',
+        purchase_units: [
+          {
+            amount: { currency_code: fee_currency.to_s, value: outstanding_fees.to_s },
+            # amount: { currency_code: "EUR", value: outstanding_fees.to_s },
+          },
+        ],
+      }
+
+      response = paypal_connection(url).post do |req|
+        req.headers['PayPal-Partner-Attribution-Id'] = AppSecrets.PAYPAL_ATTRIBUTION_CODE
+        req.headers['PayPal-Auth-Assertion'] = get_paypal_auth_assertion(registration.competition)
+
+        req.body = payload
+      end
+
+      response.body
+    end
+  end
+
+  def self.capture_payment(competition, order_id)
+    unless Rails.env.production?
+      url = "#{EnvConfig.PAYPAL_BASE_URL}/v2/checkout/orders/#{order_id}/capture"
+
+      response = paypal_connection(url).post do |req|
+        req.headers['PayPal-Partner-Attribution-Id'] = AppSecrets.PAYPAL_ATTRIBUTION_CODE
+        req.headers['PayPal-Auth-Assertion'] = get_paypal_auth_assertion(competition)
+      end
+
+      response.body
+    end
+  end
+
+  class << self
+    def paypal_connection(url)
+      unless Rails.env.production?
+        Faraday.new(
+          url: url,
+          headers: {
+            'Authorization' => "Bearer #{generate_access_token}",
+            'Content-Type' => 'application/json',
+          },
+        ) do |builder|
+          # Sets headers and parses jsons automatically
+          builder.request :json
+          builder.response :json
+
+          # Raises an error on 4xx and 5xx responses.
+          builder.response :raise_error
+
+          # Logs requests and responses.
+          # By default, it only logs the request method and URL, and the request/response headers.
+          builder.response :logger
+        end
+      end
+    end
+  end
+
+  private_class_method def self.generate_access_token
```

Comment:
nice, didn't know this existed. although I'm not sure we need to be _this_ cautious about it.
Not really a blocker against merging this PR, but a bit food for thought.

Reply:
Okay, makes sense for my inner coding mantra but doesn't make sense in terms of "I feel like we're using this feature (aka `private_class_method`) in exactly zero other places throughout our monolith codebase)

Reply:
Would you see that as a reason to not adopt it going forward? I understand concerns re consistency/styling, but given that it declares pretty clearly what it does and improves the readability I'd be in favour of it even if it's new relative to our codebase